app/toolbox-server/Toolbox/Render/ModuleGraph.hs

```diff
@@ -112,3 +157,73 @@ renderModuleGraph ss =
             []
             [ pre [] [text $ formatModuleGraphInfo (sessionModuleGraph sessionInfo)]
             ]
+
+newGA :: Graph -> IO GraphAttributes
+newGA g = newGraphAttributes g (nodeGraphics .|. edgeGraphics .|. nodeLabel .|. nodeStyle)
+
+makeReducedGraphReversedFromModuleGraph :: ModuleGraphInfo -> [(Int, [Int])]
+makeReducedGraphReversedFromModuleGraph mgi =
+  let nVtx = length $ mginfoModuleNameMap mgi
+      es = makeEdges $ mginfoModuleDep mgi
+      g = buildG (1, nVtx) es
+      seeds = filterOutSmallNodes mgi
+      tordVtxs = topSort g
+      tordSeeds = filter (`elem` seeds) tordVtxs
+      reducedGraph = makeReducedGraph g tordSeeds
+   in IM.toList $ mkRevDep reducedGraph
+
+layOutGraph :: [(Int, Text)] -> [(Int, [Int])] -> IO GraphVisInfo
+layOutGraph nameMap graph = do
+  bracket newGraph delete $ \g ->
+    bracket (newGA g) delete $ \ga -> do
+      moduleNodeMap <-
+        IM.fromList . concat
+          <$> ( forM graph $ \(i, _) -> do
+                  case L.lookup i nameMap of
+                    Nothing -> pure []
+                    Just name -> do
+                      let width = 8 * T.length name
+                      node <- newGraphNodeWithSize (g, ga) (width, 15)
+                      appendText ga node name
+                      pure [(i, node)]
+              )
+      moduleNodeIndexList <-
+        F.toList
+          <$> IM.traverseWithKey
+            (\i node -> (i,) . fromIntegral @_ @Int <$> nodeElement_index node)
+            moduleNodeMap
+      let moduleNodeIndexRevList = fmap swap moduleNodeIndexList
+      for_ graph $ \(i, js) ->
+        for_ (IM.lookup i moduleNodeMap) $ \node_i -> do
+          let node_js = mapMaybe (\j -> IM.lookup j moduleNodeMap) js
+          for_ node_js $ \node_j ->
+            graph_newEdge g node_i node_j
+
+      bracket newSugiyamaLayout delete $ \sl -> do
+        orank <- newOptimalRanking
+        sugiyamaLayout_setRanking sl orank
+        mh <- newMedianHeuristic
+        sugiyamaLayout_setCrossMin sl mh
+        ohl <- newOptimalHierarchyLayout
+        optimalHierarchyLayout_layerDistance ohl 5.0
+        optimalHierarchyLayout_nodeDistance ohl 1.0
+        optimalHierarchyLayout_weightBalancing ohl 0.5
+        sugiyamaLayout_setLayout sl ohl
+        call sl ga
+
+        drect <- boundingBox ga
+        canvasWidth :: Double <- realToFrac <$> dRect_width drect
+        canvasHeight :: Double <- realToFrac <$> dRect_height drect
+
+        nodeLayout0 <- getAllNodeLayout g ga
+        let nodeLayout = mapMaybe replace nodeLayout0
+              where
+                replace (j, x, y, w, h) = do
+                  i <- L.lookup j moduleNodeIndexRevList
+                  name <- L.lookup i nameMap
+                  pure (name, x, y, w, h)
```

Comment:
this does seem like it'd be better as `IM.lookup` - especially on `nameMap`, we're doing two loops and doing a `lookup` on the list for each loop.

Reply:
heh. yeah. I kept using lists at the moment for prototyping (as it didn't hit performance problem yet) and was postponing to replace that globally. I think now is a good time to do that action. Thanks for the change request!

Reply:
Now converted to IntMap #20